fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ReplaceExpressionByChildOutput.java

```diff
@@ -0,0 +1,99 @@
+import org.apache.doris.nereids.trees.plans.logical.LogicalSort;
+import org.apache.doris.nereids.util.ExpressionUtils;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Maps;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+/**
+ * replace.
+ */
+public class ReplaceExpressionByChildOutput implements AnalysisRuleFactory {
+    @Override
+    public List<Rule> buildRules() {
+        return ImmutableList.<Rule>builder()
+                .add(RuleType.REPLACE_SORT_EXPRESSION_BY_CHILD_OUTPUT.build(
+                        logicalSort(logicalProject()).then(sort -> {
+                            LogicalProject<GroupPlan> project = sort.child();
+                            Map<Expression, Slot> sMap = Maps.newHashMap();
+                            project.getProjects().stream()
+                                    .filter(Alias.class::isInstance)
+                                    .map(Alias.class::cast)
+                                    .forEach(p -> sMap.put(p.child(), p.toSlot()));
+                            return replaceSortExpression(sort, sMap);
+                        })
+                ))
+                .add(RuleType.REPLACE_SORT_EXPRESSION_BY_CHILD_OUTPUT.build(
+                        logicalSort(logicalAggregate()).then(sort -> {
+                            LogicalAggregate<GroupPlan> aggregate = sort.child();
+                            Map<Expression, Slot> sMap = Maps.newHashMap();
+                            aggregate.getOutputExpressions().stream()
+                                    .filter(Alias.class::isInstance)
+                                    .map(Alias.class::cast)
+                                    .forEach(p -> sMap.put(p.child(), p.toSlot()));
+                            return replaceSortExpression(sort, sMap);
+                        })
+                )).add(RuleType.REPLACE_SORT_EXPRESSION_BY_CHILD_OUTPUT.build(
+                        logicalSort(logicalHaving(logicalAggregate())).then(sort -> {
+                            LogicalAggregate<GroupPlan> aggregate = sort.child().child();
+                            Map<Expression, Slot> sMap = Maps.newHashMap();
+                            aggregate.getOutputExpressions().stream()
+                                    .filter(Alias.class::isInstance)
+                                    .map(Alias.class::cast)
+                                    .forEach(p -> sMap.put(p.child(), p.toSlot()));
+                            return replaceSortExpression(sort, sMap);
+                        })
+                ))
+                .build();
+    }
+
+    private LogicalPlan replaceSortExpression(LogicalSort<? extends LogicalPlan> sort, Map<Expression, Slot> sMap) {
+        List<OrderKey> orderKeys = sort.getOrderKeys();
+        AtomicBoolean changed = new AtomicBoolean(false);
```

Comment:
Why use `AtomicBoolean`? Does some race conditions exist?

Reply:
lambda in stream api cannot use normal local variable